Rubberduck.VBEEditor/ComManagement/TypeLibs/TypeLibs.cs

```diff
@@ -0,0 +1,1162 @@
+﻿using System;
+using System.Runtime.InteropServices;
+using System.Collections;
+using System.Collections.Generic;
+using System.Text;
+using Rubberduck.VBEditor.SafeComWrappers.Abstract;
+using Rubberduck.VBEditor.ComManagement.TypeLibsAbstract;
+using ComTypes = System.Runtime.InteropServices.ComTypes;
+using Reflection = System.Reflection;
+
+
+
+
+// TODO expose references collection
+// TODO comments/XML doc
+// TODO a few FIXMEs
+// TODO add DoesImplement() support for "typeLibName.interfaceName" format, and arrays
+
+
+
+
+// USAGE GUIDE:   see class VBETypeLibsAPI for demonstrations of usage.
+//
+// The root object for exposure of the type libraries is VBETypeLibsAccessor.  It takes the VBE in its construtor.
+// The main wrappers (TypeLibWrapper and TypeInfoWrapper) can be used as regular ITypeLib and ITypeInfo objects through casting.
+//
+// THIS IS A WORK IN PROGRESS.  ERROR HANDLING NEEDS WORK, AS DOES A FEW OF THE HELPER ROUTINES.
+//
+// WARNING: when using VBETypeLibsAccessor directly, do not cache it  
+//   The type library is LIVE information, so consider it a snapshot at that very moment when you are dealing with it
+//   Make sure you call VBETypeLibsAccessor.Dispose() as soon as you have done what you need to do with it.
+//   Once control returns back to the VBE, you must assume that all the ITypeLib/ITypeInfo pointers are now invalid.
+//
+// CURRENT LIMITATIONS:
+// At the moment, enums and UDTs are not exposed through the type libraries
+// Constants names are not available
+
+// IMPLEMENTATION DETAILS:
+// There are two significant bugs in the VBE typeinfos implementations that we have to work around.
+// 1)  Some implementations of ITypeInfo provided by the VBE will crash with an AV if you call IProvideClassInfo::GetClassInfo on them.
+//      And guess what method the CLR calls on all COM interop objects when creating a RCW?  You guessed it.
+//      So, we use an aggregation object, plus ITypeInfo and ITypeLib wrappers to circumvent this VBE bug.
+//
+// 2)  The ITypeInfo for base classes of UserForms crash with an AV if you call ITypeInfo::GetDocumentation(MEMBERID_NIL) to get the type name
+//     We've got to remember that the VBE didn't ever intend for us to get hold of these objects, so there will be little bugs.
+//     This bug is also resolved in the provided wrappers.
+//
+// All the extended functionality is exposed through the wrappers.
+
+namespace Rubberduck.VBEditor.ComManagement.TypeLibs
+{
+    public class StringLineBuilder
+    {
+        StringBuilder _document = new StringBuilder();
+
+        public override string ToString() => _document.ToString();
+
+        public void AppendLine(string value = "")
+            => _document.Append(value + "\r\n");
+
+        public void AppendLineNoNullChars(string value)
+            => AppendLine(value.Replace("\0", string.Empty));
+    }
+
+    public class StructHelper
```

Comment:
yup, very true, thanks

Reply:
@bclothier no need to instantiate a class to invoke its `static` members, but I agree if all members are static, the class itself should be static *so it can't be instantiated* ;-)